docs/RFCS/sql_batching.md

```diff
@@ -0,0 +1,669 @@
+to extend for statement batching. A user would simply activate statement batching
+through some kind of [setting](#opt-in-vs-always-on) and would then send groups of
+statements separated by semicolons in a single request. Our SQL engine would detect
+statements that could be run in parallel and would do so whenever possible. It would
+then return a single response with multiple result sets, like it already does for grouped
+statements like these. A list of pros and cons to this approach are listed below:
+
+- [+] Drivers are already expecting multiple result sets all at once when they send
+  semicolon-separated statements
+- [+] Never needs to mock out statement results
+- [+] Allows [error handling](#error-handling) to remain relatively straightforward
+  because the executor has all statements on hand at the time of batch execution and
+  will not prematurely testify to the success or failure of any statement before its
+  execution
+- [+] Fits a nice mental model/easy to reason about
+- [+] Does not need to be opt-in, although we probably still want it to be, at least
+  at first
+- [+] More generally applicable
+- [+] Could support batched reads
+- [-] May not be possible to use with ORMs (or may at least require custom handling)
+- [-] Inflexible with regard to allowing users to decide which statements get executed
+  in parallel
+
+### Conversational API Support
+
+In contrast to sending a list of statements separated by semicolons, most SQL clients
+talk to databases using a single statement at-a-time conversational API. This means that
+the clients send a statement and wait for the response before proceeding. They expect at the
+point of a response for a statement that the statement has already been applied. An
+alternate approach to batching statements for semicolon-separated lists is to build
+batches between statement declarations when clients use this more standard conversational
+API. To do so, we would need to buffer statements instead of executing them immediately
+when we decide (or are told) to batch statements. We would then mock out a successful response
+value to immediately send back to the client. After this, the `sql.Executor` would wait
+for the next statement in the conversation to decide whether to continue building up a
+collection of statements or to execute all pending statements. The decision to either execute
+the accumulated statements or add to the batch would come down to if the new statement was
+[independent](#independent-statements) from the previous set of statements. Additionally,
+we would always need to execute the pending batch on read queries and transaction commit
+statements.
+
+To perform statement batching using this interface, users would need to specify which
+statements could be batched. They would do so using some new syntax, for instance appending
+`RETURNING NOTHING` to the end of INSERT and UPDATE statements. In turn, adding this
```

Comment:
As it stands right now, the proposal is not planning on adding any new session variables. `RETURNING NOTHING` would be the only user-facing syntax change.